[FIX] fix NoseTester's raise_warning default

njsmith opened the pull request:
Our test-runner's raise_warning mode traditionally has varied depending
on whether we have a development or release version of numpy: for
development versions we raise on warnings, and for release versions we
don't. This is all very sensible... _if_ you're running numpy's test
suite. But our test-runner is also used by other packages like scipy,
and it doesn't make sense for scipy's raise_warning mode to vary
depending on whether _numpy_ is a development or release version. (It
should vary depending on whether the scipy-under-test is a development
or release version.) So this commit moves the numpy-version-dependent
raise_warning logic out of the generic NoseTester class and into
numpy-specific code.

(See scipy/scipy#5609 for more discussion.)

njsmith committed:
[FIX] fix NoseTester's raise_warning default

Our test-runner's raise_warning mode traditionally has varied depending
on whether we have a development or release version of numpy: for
development versions we raise on warnings, and for release versions we
don't. This is all very sensible... *if* you're running numpy's test
suite. But our test-runner is also used by other packages like scipy,
and it doesn't make sense for scipy's raise_warning mode to vary
depending on whether *numpy* is a development or release version. (It
should vary depending on whether the scipy-under-test is a development
or release version.) So this commit moves the numpy-version-dependent
raise_warning logic out of the generic NoseTester class and into
numpy-specific code.

(See scipy/scipy#5609 for more discussion.)

Eric89GXL commented on numpy/__init__.py:
```diff
@@ -185,8 +185,12 @@ def pkgload(*packages, **options):
+        bench = Tester(raise_warnings="develop").bench
+    else:
+        test = Tester(raise_warnings="release").test
+        bench = Tester(raise_warnings="release").bench
```
Minor nitpick, but this could be a tiny bit more DRY as something like:
```
if '.dev0' in __version__:
    raise_warnings = 'develop'
else:
    raise_warnings = 'release'
tester = Tester(raise_warnings=raise_warnings)
test = tester.test
bench = tester.bench
```
Separate instances of `Tester` can still be used to remain DRY if it's necessary, but is it necessary for some reason? (If so a comment would be helpful because it left me wondering.)

njsmith replied:
This is `numpy/__init__.py`, so your sample code is missing several `del` statements that are necessary to avoid namespace pollution (and even without that, it's already longer than the code as written). I tried writing it a few ways, and this seemed the most readable overall.

AFAIK there's no reason for the `test` and `bench` object instances to be different; in fact AFAIK there's no reason for `bench` to exist at all, b/c I think we use asv these days? But here I just went for the minimal change to keep the old behavior...

Eric89GXL replied:
Ahh sorry, didn't notice the location.

FYI if line count is at a premium, my proposed 4-line conditional can actually fit on a single line. Add in the other three lines plus the requisite new single del line yields 5 total lines to do the same thing instead of the current six (with the advantage of using a single instance). But it probably doesn't matter much either way...

njsmith replied:
It's not so much that line count is at a premium, as that the point of DRY is to keep things simple and obvious and easy to modify, so making things more complicated in order to make them DRY is a dubious win :-)

Eric89GXL commented on numpy/testing/nosetester.py:
```diff
@@ -176,7 +176,18 @@ class NoseTester(object):
+        # numpy.testing. (They want to set this based on whether *they* are a
+        # release or develop version, not whether numpy is.) So we continue to
+        # accept 'None' for back-compat, but it's now just an alias for the
+        # default "release".
```
This makes sense, except that the conditional below starting `if raise_warnings is None and` will never evaluate to `True` anymore, so I guess it can be removed.

njsmith replied:
Doh, I somehow managed to delete that code locally without it ending up in the git commit :-) Fixed.

rgommers commented:
This makes sense to me. I'm not too bothered by all the numpy submodule ``test`` callables not having this behavior, but it could be added for consistency.

njsmith committed:
[TST] Refactor new raise_warnings logic for subpackage test suites

njsmith commented:
Refactored a bit -- the new code makes the `numpy/__init__.py` logic simpler, and I went ahead and fixed the submodule test suites while I was at it.

(I am somewhat curious whether there's any reason not to just delete the submodule test suites and the `bench` functions, though it's out of scope for this PR regardless.)

rgommers commented:
> (I am somewhat curious whether there's any reason not to just delete the submodule test suites and the bench functions, though it's out of scope for this PR regardless.)

The ``bench`` functions can go now that we have ``asv``, but the submodule tests are very useful. If you're working on one submodule you can iterate much faster with ``submodule.test()`` rather than ``numpy.test()``.

njsmith commented:
Can't you just use nosetests's filtering facilities, which are much more powerful and fine-grained? e.g. `./runtests.py numpy.linalg` or `./runtests.py numpy.lib.tests.test_packbits` or `./runtests.py numpy.lib.tests.test_stride_tricks:test_same`?

njsmith commented:
Doh, this PR is currently totally broken, I'll investigate.

rgommers commented:
> Can't you just use nosetests's filtering facilities, which are much more powerful and fine-grained?

I could, but it's not what I'm used to. In some cases it's equivalent, in some cases not (think doing this for an installed numpy - where would you get ``runtests.py`` from?).

Also being able to run tests from within an interpreter and not having to be in the root dir of the repo (that's not where one should keep test scripts, so that's not where I am usually) are useful things imho. And anyway, there is no problem with keeping the ``test`` submodule feature.

njsmith commented:
If numpy is installed you can just type `nosetests numpy.linalg` or whatever I think... anyway yeah, doesn't really matter, was just an idle curiosity :-).

This PR is ready I think.

rgommers commented:
> If numpy is installed you can just type ``nosetests numpy.linalg``

Then you don't get the numpy plugins, so for example all tests marked as knownfail will give test errors.

rgommers commented on numpy/testing/nosetester.py:
```diff
@@ -514,3 +517,10 @@ def bench(self, label='fast', verbose=1, extra_argv=None):
         return nose.run(argv=argv, addplugins=add_plugins)
+
+def _numpy_tester():
+    if ".dev0" in np.__version__:
```
this will break builds under some circumstances (try ``build_ext -i`` twice in a row), you need the ``hasattr`` check that you deleted from ``NoseTester.__init__``.

njsmith replied:
Huh. I'm just going to fix this and not ask any questions.

rgommers commented:
LGTM now, merged. Thanks Nathaniel.
